tt_metal/impl/dispatch/kernels/cq_dispatch.cpp

```diff
@@ -213,9 +213,12 @@ void relay_to_next_cb(uint32_t data_ptr,
         preamble_size == 0 || preamble_size == sizeof(dispatch_packet_header_t),
         "Dispatcher preamble size must be 0 or sizeof(dispatch_packet_header_t)");
 
+    bool page_acquired = false;
     // The downstream packetizing stage will initialize the other fields, but it needs info on
     // the length of the transfer to be packetized.
     if (preamble_size > 0) {
+        cb_acquire_pages<my_noc_xy, my_downstream_cb_sem_id>(1); // XXXX optimize, take all availabl
+        page_acquired = true;
         uint64_t downstream_noc_addr = get_noc_addr_helper(downstream_noc_xy, downstream_cb_data_ptr);
```

Comment:
either need to check downstream_cb_data_ptr for being at downstream_cb_end or should assert it is not.  I'm pretty sure we need the check

Reply:
At the end of cmd handling, once we exit the transfer loop, we round up to next page:
   ```
 // Move to next page
    downstream_cb_data_ptr = round_up_pow2(downstream_cb_data_ptr, dispatch_cb_page_size);
    if (downstream_cb_data_ptr == downstream_cb_end) {
        downstream_cb_data_ptr = downstream_cb_base;
    }
```
When `relay_to_next_cb` is called, won't `downstream_cb_data_ptr` be at the start of a valid page, since last command would have put it there. And the initial value will be start of cb out of reset.

Reply:
perfect